### ISSUE TITLE
LUA events test

### ISSUE THREAD

**Issue by JimB40** [FrSkyRC/ETHOS-Feedback-Community#2066]
Hello,

Made some initial tests to check events capturing implementation in ETHOS.

Test env:
- using X10EXPRESS & X20 simulators from 1.4.0 release
- Widget in Full Screen layout, no title

Findings.

1. General
a. Widget events handling. Follows #2040. 
- LUA scripts in Widget mode get key or touch events only when focused. 
- Widget focused state is marked by underlying highlight. 
- To get widget focused you need to use ROTARY or short press ENTER. 
- After some predefined time (2-3 seconds) widget focus is lost.

I was quite confused at first why events are recieved then stop until read #2040 and found about focus.
In case of LUA widget being displayed using full screen layout things get even more complicated. Drawing something that covers entire screen causes that focus mark is invisible. So user don't know in what state widget is.
I agree with Udo that in case widget is LUA script all key events are passed to LUA event handler. Except 35 KEY_EXIT_LONG that can ovverride LUA event handler and pass events handlig back to system.Additionally system.exit() function can give LUA developer possibility to code script closing (or passing events handling in case of Widget) back to OS.
In terms of how to enable LUA event handler IMHO it should be registerWidget option (eventHandler=true/false) to prevent lock-up by not so well written LUA scritpt. 

b. SystemTool mode.
So far I can't find a way to run LUA script in full-screen (BottomBar with icons disapears but ToolBar stays visible). IMHO if that was intended should be changed to option (topBar=true/false), as KEY_EXIT_LONG will always 'close' LUA SystemTool

2. X10ESPRESS SystemTool mode testing.
a. for all keys no 'KEY_XXX_REPT' constants are defined even 64-69 events are recieved
b. 4099, 4100 (ROTARY_LEFT, ROTARY_RIGHT) event are not recieved at all
c. 33 (KEY_ENTER_BREAK) is not recieved
d. after recieving 35 (KEY_EXIT_LONG) - all key inputs are suspended. No events recieved anymore. I suppose SystemTool (LUA script) should be exited.

3. X10ESPRESS Widget mode.
a. for all keys no 'KEY_XXX_REPT' constants are defined even 64-69 events are recieved
b. events are recieved only for certain time when widget is focused (1)

4. X20 SystemTool mode testing.
a. for all keys no 'KEY_XXX_REPT' constants are defined even 64-69 events are recieved
b. 4099, 4100 (ROTARY_LEFT, ROTARY_RIGHT) event are not recieved at all
c. after recieving 35 (KEY_ENTER_LONG) - all key inputs are suspended. No events recieved anymore. I suppose SystemTool (LUA script) should be exited.
d. TOUCH events are not recieved at all

5. X20 Widget mode
a. for all keys no 'KEY_XXX_REPT' constants are defined even 64-69 events are recieved
b. even small sliding after 16640 (TOUCH_START) is recieved will start OS event handling to change screen causing jitter.
c. events are recieved only for certain time when widget is focused (1)

@bsongis I don't have yet hardware so I can't test if those thimgs are related to simulator only or both simulator and firmware.
Also dont't know which points are intentionally implemented and which ones are just bugs, so appreciate comment. 

Enclosing testing LUA script 
[ETHOS_LUA_events_test.zip](https://github.com/FrSkyRC/ETHOS-Feedback-Community/files/9969479/ETHOS_LUA_events_test.zip)

**Comment by JimB40** [comment]
Made another perforarmance test on X20 simu
1. I've added lcd.invalidate() in LUA event handler function ->  event handling speed degraded so less KEY_XXX_LONG events are recieved
2. I've added lcd.invalidate() in LUA wakeup function ->  event handling speed degraded so no KEY_XXX_LONG events are recieved. Only KEY_XXX_FIRST and KEY_XXX_BREAK
![Screenshot 2022-11-11 at 09 17 08](https://user-images.githubusercontent.com/46420768/201299650-cdad0d93-87d1-4d91-b61a-309a50f7ff97.png)

Is it LUA event handling implemented in a way that screen redrawing affects it's performance?

**Comment by strgaltdel** [comment]
.. "keepalive" comment

#2040 can remain closed because this one includes the issue / solution
it won't be fixed with the 1.5.0 release which is underway, right ?

**Comment by azaz44** [comment]
I faced similar problems, but I was testing with `registerSystemTool` instead of widgets. My findings
- X20 simulator does not get rotary right / left at all
- X10 express simulator does get ROTARY RIGHT, but does not get LEFT
- real X10 (non-express) also does get ROTARY RIGHT, but not LEFT

I'd say, for LUA developments these are critical issues.
Simple script to reproduce attached.

```
local function create()
end

local function wakeup()
end

local function event(widget, category, value, x, y)
  print("Event received:", category, value, x, y)
end

local function paint()
end

local icon = lcd.loadMask("snake.png")

local function init()
  system.registerSystemTool({name="Test", icon=icon, create=create, wakeup=wakeup, event=event, paint=paint})
end

return {init=init}
```

[test.zip](https://github.com/FrSkyRC/ETHOS-Feedback-Community/files/10324473/test.zip)

**Comment by RealTadango** [comment]
No rotary events, no touch events..... come one guys, we need this :)

**Comment by JimB40** [comment]
bump

**Comment by bsongis-frsky** [comment]
@azaz44 the `create` function creates a form inside the tool and this form will capture all rotary encoder events. I wonder what you are trying to do, but perhaps your `create` function is not needed?

**Comment by bsongis-frsky** [comment]
Touch Events added to the next nightlies

**Comment by bsongis-frsky** [comment]
@JimB40 it will be difficult to check the performance in simulator, better use the radio for that!

**Comment by bsongis-frsky** [comment]
@strgaltdel may I assume that you have understood that if your event handler returns `true`, then the default Ethos handler won't be called?

**Comment by azaz44** [comment]
> @azaz44 the `create` function creates a form inside the tool and this form will capture all rotary encoder events. I wonder what you are trying to do, but perhaps your `create` function is not needed?

I think I need to retest this, this was many many versions ago.

**Comment by bsongis-frsky** [comment]
It will be in next nightlies that touch events are forwarded, my commit has been pushed after the start of the nightlies

**Comment by bsongis-frsky** [comment]
I clear the milestone as I am not sure something goes bad ;)

**Comment by JimB40** [comment]
printed also on console via serial port and value=0.0

**Comment by JimB40** [comment]
Can be closed with 1.5 release
So far havent spotted issues.
